## About

This project focuses on predicting the selling price of used cars using machine learning. The dataset contains information about cars such as car name, year, present price, kilometers driven, fuel type, seller type, transmission, and previous owner information.

A Random Forest Regression model is used to learn the relationship between these features and the car's selling price. The model's performance is evaluated using MAE, MSE, RMSE, and R² Score.


## Objective

The main objective of this project is to build a machine learning regression model that can predict the selling price of a used car based on its available features.

The project aims to:

* Perform data exploration and preprocessing.
* Prepare categorical and numerical features for machine learning.
* Train a Random Forest Regression model.
* Predict car selling prices.
* Evaluate model performance using MAE, MSE, RMSE, and R² Score.
* Analyze the difference between actual and predicted prices.


## Table of Contents

1. About the Dataset
2. Objective
3. Importing Libraries
4. Loading the Dataset
5. Understanding the Dataset
6. Data Information
7. Checking Missing Values
8. Checking Duplicate Values
9. Exploratory Data Analysis (EDA)
10. Feature Selection
11. Data Preprocessing
12. Encoding Categorical Variables
13. Defining Features and Target
14. Train-Test Split
15. Random Forest Regression
16. Model Training
17. Model Prediction
18. Model Evaluation
19. Actual vs Predicted Values
20. Error Analysis
21. Conclusion


In [10]:
import pandas as pd

In [11]:
df = pd.read_csv("car data.csv")

In [12]:
df

,Car_Name,Year,Selling_Price,Present_Price,Kms_Driven,Fuel_Type,Seller_Type,Transmission,Owner
0,ritz,2014,3.35,5.59,27000,Petrol,Dealer,Manual,0
1,sx4,2013,4.75,9.54,43000,Diesel,Dealer,Manual,0
2,ciaz,2017,7.25,9.85,6900,Petrol,Dealer,Manual,0
3,wagon r,2011,2.85,4.15,5200,Petrol,Dealer,Manual,0
4,swift,2014,4.60,6.87,42450,Diesel,Dealer,Manual,0
...,...,...,...,...,...,...,...,...,...
296,city,2016,9.50,11.60,33988,Diesel,Dealer,Manual,0
297,brio,2015,4.00,5.90,60000,Petrol,Dealer,Manual,0
298,city,2009,3.35,11.00,87934,Petrol,Dealer,Manual,0
299,city,2017,11.50,12.50,9000,Diesel,Dealer,Manual,0


In [13]:
df.shape

(301, 9)

In [14]:
df.isna().sum()

Car_Name         0
Year             0
Selling_Price    0
Present_Price    0
Kms_Driven       0
Fuel_Type        0
Seller_Type      0
Transmission     0
Owner            0
dtype: int64

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Car_Name       301 non-null    object 
 1   Year           301 non-null    int64  
 2   Selling_Price  301 non-null    float64
 3   Present_Price  301 non-null    float64
 4   Kms_Driven     301 non-null    int64  
 5   Fuel_Type      301 non-null    object 
 6   Seller_Type    301 non-null    object 
 7   Transmission   301 non-null    object 
 8   Owner          301 non-null    int64  
dtypes: float64(2), int64(3), object(4)
memory usage: 21.3+ KB


In [16]:
df.duplicated().sum()

np.int64(2)

In [17]:
df = df.drop_duplicates()

In [18]:
df.duplicated().sum()

np.int64(0)

In [19]:
df.shape

(299, 9)

## Features and Target

The target variable for this project is `Selling_Price`, which represents the selling price of the used car.

The `Car_Name` column is removed because it is not directly used as a numerical predictive feature.

### Features (X)

* Year
* Present_Price
* Kms_Driven
* Fuel_Type
* Seller_Type
* Transmission
* Owner

### Target (y)

* Selling_Price

```python
X = df.drop(["Car_Name", "Selling_Price"], axis=1)
y = df["Selling_Price"]
```


In [54]:
x = df.drop(["Car_Name","Selling_Price"],axis=1)
y = df["Selling_Price"]

In [55]:
x

,Year,Present_Price,Kms_Driven,Fuel_Type,Seller_Type,Transmission,Owner
0,2014,5.59,27000,Petrol,Dealer,Manual,0
1,2013,9.54,43000,Diesel,Dealer,Manual,0
2,2017,9.85,6900,Petrol,Dealer,Manual,0
3,2011,4.15,5200,Petrol,Dealer,Manual,0
4,2014,6.87,42450,Diesel,Dealer,Manual,0
...,...,...,...,...,...,...,...
296,2016,11.60,33988,Diesel,Dealer,Manual,0
297,2015,5.90,60000,Petrol,Dealer,Manual,0
298,2009,11.00,87934,Petrol,Dealer,Manual,0
299,2017,12.50,9000,Diesel,Dealer,Manual,0


In [56]:
y

0       3.35
1       4.75
2       7.25
3       2.85
4       4.60
       ...  
296     9.50
297     4.00
298     3.35
299    11.50
300     5.30
Name: Selling_Price, Length: 299, dtype: float64

## Categorical Feature Encoding

The dataset contains categorical variables such as `Fuel_Type`, `Seller_Type`, and `Transmission`. Since machine learning models require numerical input, these categorical features are converted into numerical format using One-Hot Encoding.

`drop_first=True` is used to remove the first category from each feature and avoid redundant dummy variables.


In [57]:
category_cols = ["Fuel_Type","Seller_Type","Transmission"]

In [58]:
x = pd.get_dummies(x,columns=category_cols,drop_first=True,dtype=int)

In [59]:
x

,Year,Present_Price,Kms_Driven,Owner,Fuel_Type_Diesel,Fuel_Type_Petrol,Seller_Type_Individual,Transmission_Manual
0,2014,5.59,27000,0,0,1,0,1
1,2013,9.54,43000,0,1,0,0,1
2,2017,9.85,6900,0,0,1,0,1
3,2011,4.15,5200,0,0,1,0,1
4,2014,6.87,42450,0,1,0,0,1
...,...,...,...,...,...,...,...,...
296,2016,11.60,33988,0,1,0,0,1
297,2015,5.90,60000,0,0,1,0,1
298,2009,11.00,87934,0,0,1,0,1
299,2017,12.50,9000,0,1,0,0,1


In [60]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error,mean_absolute_error,r2_score,root_mean_squared_error
import numpy as np

## Train-Test Split

The dataset is divided into training and testing sets using an 80:20 ratio. The training data is used to train the Random Forest Regression model, while the testing data is used to evaluate its performance on unseen data.

`random_state=42` is used to ensure that the same data split is obtained each time the code is executed.


In [61]:
x_train,x_test, y_train, y_test = train_test_split(
    x,y,test_size = 0.20, random_state = 42
)

## Random Forest Regression Model

A Random Forest Regression model is created to predict the selling price of used cars. The model combines multiple decision trees and averages their predictions to produce the final regression output.

* `n_estimators=100` → The model uses 100 decision trees.
* `max_depth=None` → The trees can grow until the stopping conditions are met.
* `random_state=42` → Ensures reproducible results.

In [62]:
model = RandomForestRegressor(
    n_estimators = 100,
    max_depth = None,
    random_state = 42
)

## Model Training

The Random Forest Regression model is trained using the training dataset. The model learns the relationship between the input features and the car's selling price.

```python id="x4k8ps"
model.fit(X_train, y_train)
```

Here, `X_train` contains the training features and `y_train` contains the corresponding selling prices.


In [63]:
model.fit(x_train,y_train)

RandomForestRegressor(random_state=42)

## Model Prediction

The trained Random Forest Regression model is used to predict the selling prices for the test dataset. The model generates predicted values based on the features in `X_test`.


Here, `y_pred` contains the predicted selling prices for the test data.


In [64]:
y_pred = model.predict(x_test)

In [65]:
y_pred 

array([ 9.4234,  8.269 ,  0.4447,  6.9295, 17.3215,  5.347 ,  6.8545,
        1.1785,  7.8385,  0.2781, 14.7412,  0.2078,  0.2479,  0.4657,
        4.473 ,  5.3756,  0.6019,  1.153 , 21.8693,  0.5472,  0.5574,
        2.384 ,  6.324 ,  2.301 ,  5.997 ,  8.9757,  8.7848,  1.1415,
        4.7387,  5.2493,  3.409 ,  4.474 ,  5.97  ,  2.684 ,  3.207 ,
        6.4725,  1.156 ,  0.1598,  1.16  , 17.133 ,  6.8895,  9.2773,
        0.9073,  3.5207,  0.5546,  0.199 ,  9.6395,  4.8265,  3.917 ,
        0.2012,  0.5931,  0.3387, 17.3874, 10.0462,  7.6881,  6.731 ,
        3.255 ,  3.629 , 10.4409,  7.3891])

## Actual vs Predicted Values

A DataFrame is created to compare the actual selling prices from the test dataset with the prices predicted by the Random Forest Regression model. This comparison helps in understanding how closely the model's predictions match the actual values.

In [66]:
result = pd.DataFrame({
    "Actual" : y_test,
    "Prediction" : y_pred
})

In [67]:
result

,Actual,Prediction
283,8.99,9.4234
267,8.35,8.2690
166,0.45,0.4447
9,7.45,6.9295
78,5.25,17.3215
280,5.25,5.3470
95,5.85,6.8545
111,1.15,1.1785
5,9.25,7.8385
175,0.38,0.2781


## Model Evaluation

The performance of the Random Forest Regression model is evaluated using four regression metrics: Mean Absolute Error (MAE), Mean Squared Error (MSE), Root Mean Squared Error (RMSE), and R² Score.<br>

- MAE measures the average absolute difference between actual and predicted selling prices.
- MSE calculates the average squared difference between actual and predicted values.
- RMSE is the square root of MSE and indicates the typical magnitude of prediction errors.
- R² Score measures how well the model explains the variation in the target variable.

In [68]:
MSE = mean_squared_error(y_test,y_pred)
MAE =  mean_absolute_error(y_test,y_pred)
RMSE = np.sqrt(MSE)
r2 = r2_score(y_test,y_pred)

print("MAE :", MAE)
print("MSE :", MSE)
print("RMSE :", RMSE)
print("R2 Score :",r2)

MAE : 1.4663883333333343
MSE : 12.256879635833338
RMSE : 3.500982667171224
R2 Score : 0.5244342943414741


## Error Analysis

An error column is created to calculate the difference between the actual and predicted selling prices. This helps identify how much the model's prediction differs from the actual selling price for each car.

```python
result["Error"] = result["Actual"] - result["Prediction"]
```

A positive error indicates that the actual selling price is higher than the predicted price, while a negative error indicates that the predicted price is higher than the actual selling price.


In [69]:
result["Error"] = result["Actual"] - result["Prediction"]

In [70]:
result["Error"]

283    -0.4334
267     0.0810
166     0.0053
9       0.5205
78    -12.0715
280    -0.0970
95     -1.0045
111    -0.0285
5       1.4115
175     0.1019
99     -5.0912
197    -0.0478
186     0.0021
156     0.0143
58     -0.3730
61     -0.8756
149    -0.0919
110     0.0470
64     11.1307
142     0.0528
157    -0.0774
106    -1.0340
249    -1.0740
47     -1.2510
43     -0.4970
277     0.7243
282    -0.5348
118    -0.0415
215    -0.2387
238    -0.4493
18     -0.1590
241     0.2760
34     -0.4700
25      0.3160
46     -0.5570
7       0.0275
115    -0.0460
196     0.0102
113    -0.0100
94    -13.1330
76     -1.3895
83      3.2227
120     0.1427
77     -2.0207
131     0.1954
199    -0.0790
212     1.6105
290    -0.3265
221     0.5830
180     0.0988
146    -0.0431
188    -0.0887
85    -14.8874
250     2.8538
279    -1.4381
74     -1.8310
246     0.4950
26      0.5210
211     1.3091
60     -0.4391
Name: Error, dtype: float64

### Conclusion

The Random Forest Regression model was used to predict the car's Selling Price. The model achieved an MAE of 1.47, an RMSE of 3.50, and an R² score of 52.44%. The results indicate that the model provides a reasonable baseline for predicting car selling prices, although there is scope for further improvement through hyperparameter tuning and feature engineering.
